### DISCUSSION TITLE
Question | Unban after restart

### DISCUSSION THREAD

**Discussion by Keller18306** [fail2ban/fail2ban#3085] [Q&A] [answered]
Why is the list of bans cleared when restarting and reloading in fail2ban-client? How can I disable this?

**Comment by sebres** [comment, marked as answer]
> Why is the list of bans cleared when restarting and reloading in fail2ban-client?

A restart is like a stop + start, when fail2ban stops it resets everything to leave behind a clean and tidy state (as would be that fail2ban has been not started).
Especially it is important If during restart something action-related is changed (like ports, protocols, or even whole action), so fail2ban will flush (unban) old state and reban with new state (e. g. using other banning actions/parameters).
Also there can be some time between stop and restart, so some banned tickets may be outdated (and will not be banned).

For other needs there is `?sudo? fail2ban-client reload` one can use instead of restart - `reload` does not restart fail2ban process so doesn't unban active tickets, just updates current configuration (e. g. `failregex` etc).

> How can I disable this?

One cannot disable this. 
There are several possibilities to workaround this, e. g. one could either write custom action or supply `actionflush` and `norestored` to the action, where `actionflush` does nothing and due to `norestored=1`, `actionban` would ignore restored tickets by start, see #3038 for example, so something like this:
```
banaction = %(known/banaction)s[norestored=1, actionflush=true]
```
Just note that the banned entries in chains would be no more under control of fail2ban after restart, so they will never be unbanned anymore (so one would need to unban them manually using `iptables`, `ipset`, `nft` whatever.
Although one could write complex actions ignoring duplicates (for IPs left behind by previous fail2ban process), so without `norestored` and with some extra check, ignoring IPs already available in chains by reban, but...

But I don't understand why it is necessary at all or else by which scenarios it is expected?

**Reply by Keller18306** [reply]
My project is under attack. fail2ban does not have time to view all the logs and freezes. For some reason, it is at 100% load, but the logs stop going. I have to restart it.

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/129579293-6075cacf-eba6-4e7f-86b4-865e49358e68.png)
Yes, reload helps too.

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/129579491-03c34d9d-4849-40c2-b9ed-c7b8bcaf863e.png)
However, there is no more attack, but it gives a huge load.

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/129589557-997b9156-1b20-4578-986c-c6937fb0943e.png)
There is also such an alignment. The first time everything works fine. But under heavy load, it finds everything, but does not want to ban at all.

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/129590656-6a9008a1-3533-4c1e-ae68-0e7014e53f73.png)
regexp works done

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/129591807-bbce6b5d-f68a-4f44-9d97-6d67de60327d.png)
the file size grows very quickly, so fail2ban may not have time to check everything. Or fail2ban is trying to reach an end that doesn't exist.

**Reply by sebres** [reply]
> fail2ban does not have time to view all the logs and freezes. 

Never saw that fail2ban "freezes".
Fail2ban can process message slower than they are coming (due to several reasons, see below), but it is not freezing, just the amount of messages to process grows (and some IPs get banned later as it would be possible normally).

However there was a known bug or rather a certain bottleneck that can result to a "snowball" similar race condition that could retard the bans, see #2660. This was already fixed in newer versions (0.11.2 or 0.10.6).

> However, there is no more attack, but it gives a huge load.

If the banned attackers have left many messages (before they got banned), fail2ban can surely process the messages until they will end.

I don't know exactly what it is... 
Typically, the `htop` (with threads tree) shows this picture on new fail2ban version (version >= v.0.10.5 the threads have the name of filter or action):
<details><summary>output of htop for fail2ban process with thread tree ...</summary>

```
  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command
 3063 f2b        20   0  237M 25408  6400 S  1.3  0.6 28h27:03 ├─ python fail2ban-server --async -b -s /run/f2b/f2b.sock -p /run/f2b/f2b.pid --loglevel INFO ...
14922 f2b        20   0  237M 25408  6400 S  0.0  0.6  0:00.36 │  ├─ f2b/observer
 3096 f2b        20   0  237M 25408  6400 S  0.0  0.6 58:22.39 │  ├─ f2b/a.net-inval
 3095 f2b        20   0  237M 25408  6400 S  0.0  0.6 53:19.90 │  ├─ f2b/f.net-inval
 3094 f2b        20   0  237M 25408  6400 S  0.0  0.6 56:39.63 │  ├─ f2b/a.mail
 3093 f2b        20   0  237M 25408  6400 S  0.0  0.6 36:22.63 │  ├─ f2b/f.mail
 3092 f2b        20   0  237M 25408  6400 S  0.0  0.6 57:16.01 │  ├─ f2b/a.nginx-bad
 3091 f2b        20   0  237M 25408  6400 S  0.0  0.6 41:52.46 │  ├─ f2b/f.nginx-bad
 3090 f2b        20   0  237M 25408  6400 S  0.7  0.6 55:59.68 │  ├─ f2b/a.nginx-web
 3089 f2b        20   0  237M 25408  6400 S  0.0  0.6 36:41.41 │  ├─ f2b/f.nginx-web
 3088 f2b        20   0  237M 25408  6400 S  0.0  0.6 56:19.93 │  ├─ f2b/a.nginx-oth
 3087 f2b        20   0  237M 25408  6400 S  0.0  0.6 37:24.08 │  ├─ f2b/f.nginx-oth
 3086 f2b        20   0  237M 25408  6400 S  0.0  0.6 56:32.34 │  ├─ f2b/a.nginx-ctr
 3085 f2b        20   0  237M 25408  6400 S  0.0  0.6 39:27.93 │  ├─ f2b/f.nginx-ctr
 3084 f2b        20   0  237M 25408  6400 S  0.0  0.6 56:36.56 │  ├─ f2b/a.pam-gener
 3083 f2b        20   0  237M 25408  6400 S  0.0  0.6 37:22.64 │  ├─ f2b/f.pam-gener
 3082 f2b        20   0  237M 25408  6400 S  0.0  0.6  1h00:57 │  ├─ f2b/a.named-ref
 3081 f2b        20   0  237M 25408  6400 S  0.0  0.6 39:11.63 │  ├─ f2b/f.named-ref
 3080 f2b        20   0  237M 25408  6400 S  0.0  0.6 57:47.50 │  ├─ f2b/a.smtp
 3079 f2b        20   0  237M 25408  6400 S  0.0  0.6 40:05.09 │  ├─ f2b/f.smtp
 3078 f2b        20   0  237M 25408  6400 S  0.0  0.6  1h02:55 │  ├─ f2b/a.sshd
 3077 f2b        20   0  237M 25408  6400 S  0.0  0.6 37:44.50 │  ├─ f2b/f.sshd
 3064 f2b        20   0  237M 25408  6400 S  0.0  0.6  4h45:58 │  └─ f2b/observer
```
</details>

Because the thread names are speaking there, one is able to see which thread exactly is affected (has havy load).

Are they all the threads or processes in your output (e. g. some "old" fail2ban instance still running for some reason)?

Other reasons for heavy load:

1. inappropriate regexes (slow, unprecise `datepattern`, `prefregex`, `failregex` or `ignoreregex` with catch-alls, not anchored etc in filter or jail), see [wiki :: Best practice](https://github.com/fail2ban/fail2ban/wiki/Best-practice);
2. too many parasite traffic in log, see [wiki :: Best practice # Reduce parasitic log-traffic](https://github.com/fail2ban/fail2ban/wiki/Best-practice#reduce-parasitic-log-traffic);
3. fail2ban database is too large, also see https://github.com/fail2ban/fail2ban/issues/2918#issuecomment-769780199 for details;
4. logs are too large (not rotated) and `findtime` is too high, so after restart fail2ban have to process millions of entries (but normally it would calm down after several minutes);

Or a combination of all that, especially 1 & 2 together with many "bad" regexes in filter, monitoring a large log-file with few matches - note that every line which did not match will be checked against all REs, so if they are bad/slow it would surely cause high CPU load in such filter.

There may be also several other reasons, partially summarized in this https://github.com/fail2ban/fail2ban/issues/3021#issuecomment-834287295 and others. 
Just search in the closed issues.

**Reply by sebres** [reply]
> But under heavy load, it finds everything, but does not want to ban at all.

So you may be indeed affected by #2660, so either update fail2ban, or resolve the race condition by other means e. g. try to reduce count of messages (use better suitable burst or different log-file as target), speed-up regex, etc.

> the file size grows very quickly, so fail2ban may not have time to check everything. 

You may play with parameters of limit-req module (like `delay`, `burst` or `limit_req_log_level`) in order to reduce logging amount (for example log only if nginx refuses request by delay due to rate exceeding) or force to refuse it faster (but log rarer).
But I think solving the #2660 race condition may indeed help here.

> Or fail2ban is trying to reach an end that doesn't exist.

The end of file is not really interesting for fail2ban in this matter.

**Reply by Keller18306** [reply]
Yes, most likely I have the same case as that person.

My version is old
![image](https://user-images.githubusercontent.com/23178771/129596599-b36f9430-f688-4442-88c3-a94cff309a99.png)

How can I upgrade? It's just that I have everything updated in the debian packages.
![image](https://user-images.githubusercontent.com/23178771/129596841-82820310-3b1c-4409-8944-1151f4287c28.png)

**Reply by sebres** [reply]
https://github.com/fail2ban/fail2ban/wiki/How-to-install-or-upgrade-fail2ban-manually

To be sure upgrade to new version don't introduce compat issues (like different paths prefixes or python-versions), ensure to stop and uninstall original version firstly (after you make backup of your custom configs and changes in `/etc/fail2ban`).

**Reply by Keller18306** [reply]
![image](https://user-images.githubusercontent.com/23178771/130113487-c97e17f9-8e2b-4695-b540-5c93574c7d43.png)

**Reply by Keller18306** [reply]
fail2ban stopped reading error.log

**Reply by sebres** [reply]
Fail2ban can stop to process jail for 3 reasons: 
- there is an unexpected error occurred (basically should be in fail2ban.log)
- unhandled exception in filter thread (normally also logged)
- the jail was set in idle mode (for example from client)

A freeze can be caused in 0.10 by block in sqlite-database (0.11 doing it asynchronously), and it is most of the bottlenecks before 0.11.
Which version do you use now? How large is your database?
You can try the steps described in https://github.com/fail2ban/fail2ban/issues/2918#issuecomment-769780199 and below.

I don't really have the necessary info to say what is happening there or what may be wrong...
- there are 3 jails, which exactly do you mean?
- are you really sure there are still new messages in log (nginx still logging in the same location)?
some log-excerpt of both logs would be nice (e. g. `tail -f /var/log/fail2ban.log /path/to/nginx/error.log`) so one could see few of both messages together;
- was the log-file rotated? (so fail2ban doesn't recognize the rotation)
- which backend has this jail? (pyinotify or polling, I hope it was not changed to systemd, because it's for journal only)

If it's really a freeze, set DEBUG or HEAVYDEBUG levels to jail may help, but it is totally unpractical for huge log-files (because will produce to large log-output).

And last but not least, please stop to post images here (it's inconvenient to track the info and is almost marginal to disrespect)... 
Cannot you do simplest copy and paste it in comment using markdown block-format, like this (between \`\`\` and \`\`\` after new-lines)?
\`\`\`
... 
\`\`\`
